File: frontend/lib/services/keycloak/keycloak_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class KeycloakService {
  static const String baseUrl = 'https://api.client.aitecservice.com/api/keycloak';

  Future<String> login({required String username, required String password}) async {
    final response = await http.post(
      Uri.parse('$baseUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'username': username, 'password': password}),
    );
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      if (data['success'] == true && data['data'] != null) {
        return data['data']['access_token'] ?? '';
      }
      throw Exception('Échec de la connexion : ${response.body}');
    } else {
      throw Exception('Échec de la connexion : ${response.body}');
    }
  }

  Future<KeycloakUserInfo> getUserInfo(String accessToken) async {
    final response = await http.get(
      Uri.parse('$baseUrl/userinfo'),
      headers: {'Authorization': 'Bearer $accessToken'},
    );
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      if (data['success'] == true && data['data'] != null) {
        return KeycloakUserInfo.fromJson(data['data']);
      }
      throw Exception('Impossible de récupérer les infos utilisateur :  {response.body}');
    } else {
      throw Exception('Impossible de récupérer les infos utilisateur :  {response.body}');
    }
  }

  Future<void> logout(String accessToken) async {
    final response = await http.post(
      Uri.parse('$baseUrl/logout'),
      headers: {
        'Content-Type': 'application/json',
        if (accessToken.isNotEmpty) 'Authorization': 'Bearer $accessToken',
      },
    );
    if (response.statusCode != 200) {
      throw Exception('Erreur lors de la déconnexion :  {response.body}');
    }
  }

  Future<List<Map<String, String>>> fetchAccessibleCompanies(String token, List<String> userGroups) async {
    print('[fetchAccessibleCompanies] Token reçu: ' + token);
    print('[fetchAccessibleCompanies] UserGroups reçus: ' + userGroups.toString());
    final response = await http.get(
      Uri.parse('https://api-client.aitecservice.com/api/keycloak/clients-accessibles'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
    );
    print('[KeycloakService] Réponse brute: ' + response.body);
    print('[KeycloakService] Status code: ' + response.statusCode.toString());
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body)['data'];
      print('[KeycloakService] Data décodée: ' + data.toString());
      final List<Map<String, String>> companies = [];
      // Ajout du sous-groupe employé si présent
      final employeeSubGroup = userGroups.firstWhere(
        (g) => g.startsWith('/Employees/') && g != '/Employees',
        orElse: () => '',
      );
      print('[KeycloakService] employeeSubGroup: ' + employeeSubGroup);
      String? defaultEmployeeCompany;
      if (employeeSubGroup.isNotEmpty) {
        defaultEmployeeCompany = employeeSubGroup.split('/').last;
        companies.add({'id': employeeSubGroup, 'name': defaultEmployeeCompany});
        print('[KeycloakService] Ajout employé: $defaultEmployeeCompany');
      }
      // Ajout des clients et sous-groupes à plat
      final clients = _flattenClientsWithMDMFi(data['clients']);
      print('[KeycloakService] Clients filtrés MDM-Fi: ' + clients.toString());
      companies.addAll(clients);
      print('[KeycloakService] Liste finale sociétés: $companies');
      return companies;
    } else {
      print('[KeycloakService] Erreur HTTP: ${response.statusCode}');
      throw Exception('Erreur lors de la récupération des sociétés accessibles');
    }
  }

  List<Map<String, String>> _flattenClientsWithMDMFi(List clients) {
    List<Map<String, String>> result = [];
    for (final client in clients) {
      final parentName = client['name'];
      final parentLogo = client['logoURL']?.toString() ?? '';
      final parentRoles = (client['roles'] as List<dynamic>? ?? []).map((r) => r['name']?.toString() ?? '').toList();
      final subGroups = client['subGroups'] as List<dynamic>? ?? [];
      // Si le client a le rôle MDM-Fi et pas de sous-groupes, on l'ajoute
      if (subGroups.isEmpty && parentRoles.contains('MDM-Fi')) {
        result.add({'id': client['id'], 'name': parentName, 'logoURL': parentLogo});
      }
      // Sinon, on regarde les sous-groupes
      for (final sub in subGroups) {
        final subRoles = (sub['roles'] as List<dynamic>? ?? []).map((r) => r['name']?.toString() ?? '').toList();
        if (subRoles.contains('MDM-Fi')) {
          result.add({
            'id': sub['id'],
            'name': '$parentName-${sub['name']}',
            'logoURL': sub['logoURL']?.toString() ?? '',
          });
        }
      }
    }
    return result;
  }
}

class KeycloakUserInfo {
  final String name;
  final List<String> groups;
  final List<String> roles;

  KeycloakUserInfo({required this.name, required this.groups, required this.roles});

  factory KeycloakUserInfo.fromJson(Map<String, dynamic> json) {
    return KeycloakUserInfo(
      name: json['name'] ?? '',
      groups: (json['groups'] as List<dynamic>?)?.map((g) => g.toString()).toList() ?? [],
      roles: (json['roles'] as List<dynamic>?)?.map((r) => r.toString()).toList() ?? [],
    );
  }

  bool get hasMDMFi => roles.contains('MDM-Fi');
  bool get isAdmin => roles.contains('permissions-admin');
  bool get isClient => roles.contains('permissions-client');
}

File: backend/services/keycloak_service.py
# -*- coding: utf-8 -*-
import os
import jwt
import requests
from typing import Optional, Dict, List
from dotenv import load_dotenv
import logging

load_dotenv()
logger = logging.getLogger(__name__)

class KeycloakService:
    """
    Service d'authentification Keycloak pour le backend
    """
    
    def __init__(self):
        self.keycloak_url = os.getenv("KEYCLOAK_URL", "https://api.client.aitecservice.com/api/keycloak")
        self.realm = os.getenv("KEYCLOAK_REALM", "master")
        self.client_id = os.getenv("KEYCLOAK_CLIENT_ID", "mdm-fi")
        
    async def verify_token(self, token: str) -> Optional[Dict]:
        """
        Vérifie un token JWT Keycloak et retourne les informations utilisateur
        """
        try:
            # Décoder le token sans vérification de signature pour l'instant
            # En production, il faudrait vérifier avec la clé publique Keycloak
            decoded_token = jwt.decode(token, options={"verify_signature": False})
            
            logger.info(f"Token décodé pour utilisateur: {decoded_token.get('preferred_username')}")
            
            return {
                'username': decoded_token.get('preferred_username'),
                'email': decoded_token.get('email'),
                'name': decoded_token.get('name'),
                'groups': decoded_token.get('groups', []),
                'realm_access': decoded_token.get('realm_access', {}),
                'resource_access': decoded_token.get('resource_access', {}),
                'sub': decoded_token.get('sub'),
                'exp': decoded_token.get('exp')
            }
            
        except jwt.ExpiredSignatureError:
            logger.warning("Token expiré")
            return None
        except jwt.InvalidTokenError as e:
            logger.warning(f"Token invalide: {e}")
            return None
        except Exception as e:
            logger.error(f"Erreur lors de la vérification du token: {e}")
            return None
    
    async def get_user_info(self, token: str) -> Optional[Dict]:
        """
        Récupère les informations utilisateur via l'API Keycloak
        """
        try:
            headers = {
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json'
            }
            
            response = requests.get(
                f"{self.keycloak_url}/userinfo",
                headers=headers,
                timeout=10
            )
            
            if response.status_code == 200:
                data = response.json()
                if data.get('success'):
                    return data.get('data')
            
            logger.warning(f"Échec récupération userinfo: {response.status_code}")
            return None
            
        except Exception as e:
            logger.error(f"Erreur lors de la récupération des infos utilisateur: {e}")
            return None
    
    def get_user_permissions(self, user_info: Dict) -> Dict[str, bool]:
        """
        Extrait les permissions utilisateur à partir des infos Keycloak
        """
        permissions = {
            'is_admin': False,
            'is_comptable': False,
            'is_direction': False,
            'can_force_check': False,
            'can_broadcast': False,
            'has_mdm_fi': False
        }
        
        try:
            # Vérifier les rôles realm
            realm_roles = user_info.get('realm_access', {}).get('roles', [])
            
            # Vérifier les rôles de ressource
            resource_roles = []
            resource_access = user_info.get('resource_access', {})
            for resource, access in resource_access.items():
                resource_roles.extend(access.get('roles', []))
            
            all_roles = realm_roles + resource_roles
            
            # Déterminer les permissions
            permissions['has_mdm_fi'] = 'MDM-Fi' in all_roles
            permissions['is_admin'] = 'permissions-admin' in all_roles
            permissions['is_comptable'] = any('comptab' in role.lower() for role in all_roles)
            permissions['is_direction'] = any('direction' in role.lower() for role in all_roles)
            
            # Permissions dérivées
            permissions['can_force_check'] = permissions['is_admin'] or permissions['is_direction'] or permissions['is_comptable']
            permissions['can_broadcast'] = permissions['is_admin']
            
            logger.info(f"Permissions calculées pour {user_info.get('preferred_username')}: {permissions}")
            
        except Exception as e:
            logger.error(f"Erreur lors du calcul des permissions: {e}")
        
        return permissions
    
    def get_client_type_from_permissions(self, permissions: Dict[str, bool]) -> str:
        """
        Détermine le type de client à partir des permissions
        """
        if permissions.get('is_admin'):
            return 'dashboard'  # Admin a accès complet
        elif permissions.get('is_direction'):
            return 'direction'
        elif permissions.get('is_comptable'):
            return 'comptable'
        else:
            return 'mobile'  # Client basique
    
    async def authenticate_websocket(self, token: str) -> Optional[Dict]:
        """
        Authentifie un token pour une connexion WebSocket et retourne les infos complètes
        """
        if not token:
            return None
        
        # Nettoyer le token (enlever "Bearer " si présent)
        if token.startswith('Bearer '):
            token = token[7:]
        
        # Vérifier le token
        user_info = await self.verify_token(token)
        if not user_info:
            # Fallback: essayer via l'API
            user_info = await self.get_user_info(token)
        
        if not user_info:
            return None
        
        # Calculer les permissions
        permissions = self.get_user_permissions(user_info)
        
        # Vérifier que l'utilisateur a accès à MDM-Fi
        if not permissions.get('has_mdm_fi'):
            logger.warning(f"Utilisateur {user_info.get('preferred_username')} n'a pas accès à MDM-Fi")
            return None
        
        # Déterminer le type de client
        client_type = self.get_client_type_from_permissions(permissions)
        
        return {
            'user_info': user_info,
            'permissions': permissions,
            'client_type': client_type,
            'username': user_info.get('preferred_username'),
            'email': user_info.get('email')
        }

# Instance globale
keycloak_service = KeycloakService()
